src/GraphQL/Types/Scalars/DecimalGraphType.cs

```diff
@@ -7,21 +9,66 @@ namespace GraphQL.Types
     /// The Decimal scalar graph type represents a decimal value.
     /// By default <see cref="GraphTypeTypeRegistry"/> maps all <see cref="decimal"/> .NET values to this scalar graph type.
     /// </summary>
-    public class DecimalGraphType : ScalarGraphType
+    public class DecimalGraphType : ScalarGraphType, ICanParseScalar
     {
+        /// <inheritdoc/>
+        public bool CanParseLiteral(IValue value)
+        {
+            return value switch
+            {
+                DecimalValue _ => true,
+                IntValue _ => true,
+                LongValue _ => true,
+                StringValue s => decimal.TryParse(s.Value, NumberStyles.Float, NumberFormatInfo.InvariantInfo, out _),
+                FloatValue f => Ret(checked((decimal)f.Value)),
+                BigIntValue b => Ret(checked((decimal)b.Value)),
+                _ => false
+            };
+
+            static bool Ret(decimal _) => true;
+        }
+
+        /// <inheritdoc/>
+        public bool CanParseValue(object value)
+        {
+            return value switch
+            {
+                decimal _ => true,
+                int _ => true,
+                long _ => true,
+                string s => decimal.TryParse(s, NumberStyles.Float, NumberFormatInfo.InvariantInfo, out _),
+                float f => Ret(checked((decimal)f)),
+                BigInteger b => Ret(checked((decimal)b)),
+                double d => Ret(checked((decimal)d)),
```

Comment:
In principle, yes, but nevertheless we cannot guarantee this. At least for a custom scalar for sure. That's why I use `try/catch` in `IsValidLiteralValue`. After all, the reason for this bug is the fact that the scalar throws an exception. This is our scalar. What can we say about custom scalars, even if our code is error prone?

I **would** write a code here that does not throw exceptions, but I do not know how 😄 .

Reply:
Just put it in a try catch block.

```cs
float f => CanConvertFloat(f),

private bool CanConvertFloat(float f)
{
    try {
        _ = checked((decimal)f);
        return true;
    } catch {
        return false;
    }
}
```

Reply:
> In principle, yes, but nevertheless we cannot guarantee this. At least for a custom scalar for sure. That's why I use `try/catch` in `IsValidLiteralValue`. After all, the reason for this bug is the fact that the scalar throws an exception. This is our scalar. What can we say about custom scalars, even if our code is error prone?

The default behavior of `CanConvertLiteral` will catch errors.  If someone overrides this with their own buggy implementation, then it is a server error, not an input error.

Reply:
Honestly we do not need to override `CanParseLiteral` for any of the classes.  But it will be a faster implementation if we do since there is one less boxing of integers parsed as decimals.

Reply:
The proposal is understandable but will require more code jumping around `try/catch`. Also, even in the case of `their own buggy implementation`, I prefer to give a meaningful error. This is the essence of this PR. I mean that `try/catch` in `IsValidLiteralValue` should stay.

Reply:
> But it will be a faster implementation if we do since there is one less boxing of integers parsed as decimals.

Of course. Think about larger structures - `Guid`, some custom scalars.

Reply:
I don't think `CanParseLiteral` should throw an error.  Nowhere in any .NET library would you find that a `TryParse` / `CanParse` implementation is allowed to throw an error.

Given that, if you want a duplicate try/catch in `IsValidLiteralValue`, I fail to see the reason why.  In case someone else writes a custom scalar with a buggy `CanParseLiteral` implementation?  I don't think it is necessary, but okay.  That doesn't mean we should write our other `CanParseLiteral` implemenations buggy knowning that in OUR consumer of the method it doesn't matter.

Reply:
Then in `DecimalGraphType` I should write 2 or even 3 methods like this one:
```
private bool CanConvertFloat(float f)
{
    try {
        _ = checked((decimal)f);
        return true;
    } catch {
        return false;
    }
}
```

for `float`, `BigInteger`, `double` . Likewise in other built-in scalars.

> Given that, if you want a duplicate try/catch in IsValidLiteralValue, I fail to see the reason why. 

In order not to do the unnecessary work of writing `try/catch` blocks/methods as showed above in all scalars.

Reply:
Well here's another option:

```cs
public class ScalarGraphType
{
    public bool CanParseLiteral(IValue value)
    {
        try {
             return OnCanParseLiteral(value);
        } catch {
             return false;
        }
    }

    protected virtual OnCanParseLiteral(IValue value)
    {
        return ParseLiteral(value) != null;
    }
}
```

Now `CanParseLiteral` will never throw an exception and you don't have to do all that "unnecessary work".

Reply:
done